File: Aigilas/Aigilas/Reactions/ComboMeter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Aigilas.Creatures;
using Aigilas.Entities;
using SPX.Text;

namespace Aigilas.Reactions
{
    public class ComboMeter
    {
        private static Dictionary<int, int> __reactions = new Dictionary<int, int>()
        {
            {12,ReactionId.SWEAT},
            {13,ReactionId.MAGMA},
            {14,ReactionId.EXPLOSION},
            {15,ReactionId.SCORCH},
            {16,ReactionId.BLIND},
            {17,ReactionId.LACTIC_ACID},
            {18,ReactionId.MIND_BLOWN},
            {23,ReactionId.VENT},
            {24,ReactionId.DROWN},
            {25,ReactionId.REFLECT},
            {26,ReactionId.DRENCH},
            {27,ReactionId.PNEUMONIA},
            {28,ReactionId.LOBOTOMY},
            {34,ReactionId.RUST},
            {35,ReactionId.PURIFY},
            {36,ReactionId.ECLIPSE},
            {37,ReactionId.RESPECT},
            {38,ReactionId.CRAFTSMAN},
            {45,ReactionId.FLASH},
            {46,ReactionId.METABOLISM},
            {47,ReactionId.FAST_FORWARD},
            {48,ReactionId.BLANK},
            {56,ReactionId.YIN_YANG},
            {57,ReactionId.EXPOSE},
            {58,ReactionId.ENLIGHTEN},
            {67,ReactionId.ATROPHY},
            {68,ReactionId.NEUROSIS},
            {78,ReactionId.CONFUSE}
        };

        private ICreature _parent;
        private List<int> _elements = new List<int>();
        private List<ComboMarker> _markers = new List<ComboMarker>();
        private const int _maxTimer = 120;
        private int _reactionTimer = _maxTimer;
        

        public ComboMeter(ICreature parent)
        {
            _parent = parent;
        }

        private void ResetComboDisplay()
        {
            _reactionTimer = _maxTimer;
            for (int ii = 0; ii < _markers.Count(); ii++)
            {
                _markers[ii].SetInactive();
            }
            _markers.Clear();
            int jj = 0;
            foreach (var element in _elements)
            {
                _markers.Add(new ComboMarker(_parent,element,jj));
                jj++;
                _markers.Last().LoadContent();
            }
        }

        public void Draw()
        {
            foreach (var marker in _markers)
            {
                marker.Draw();
            }
        }

        public void Add(int element)
        {
            if (!_elements.Contains(element))
            {
                if (_elements.Count() == 2)
                {
                    if (_elements[0] > element)
                    {
                        _elements.Insert(0, element);
                    }
                    else if (_elements[1] > element)
                    {
                        _elements.Insert(1, element);
                    }
                    else
                    {
                        _elements.Add(element);
                    }
                }
                else if (_elements.Count() == 1)
                {
                    if (_elements[0] > element)
                    {
                        _elements.Insert(0, element);
                    }
                    else
                    {
                        _elements.Add(element);
                    }
                }
                if (_elements.Count() == 0)
                {
                    _elements.Add(element);
                }
                ResetComboDisplay();
            }
        }

        private IReaction reaction;
        public void Update()
        {
            foreach (var marker in _markers)
            {
                marker.Update();
            }
            int key = 0;
            if (_elements.Count() == 3)
            {
                key = (int)_elements[0] * 100 + (int)_elements[1] * 10 + (int)_elements[2];
                React(key);
            }
            if (_elements.Count() == 2)
            {
                key = (int)_elements[0] * 10 + (int)_elements[1];
                React(key);
            }
            if (_elements.Count() == 1)
            {
                React((int)_elements[0]);
            }
        }
        private void React(int reactionId)
        {
            _reactionTimer--;
            if (_reactionTimer <= 0)
            {
                if (__reactions.Keys.Contains(reactionId))
                {
                    reaction = ReactionFactory.Create(__reactions[reactionId]);
                    if (reaction != null)
                    {
                        reaction.Affect(_parent);
                        TextManager.Add(new ActionText(ReactionId.Names[__reactions[reactionId]], 10, (int)_parent.GetLocation().PosX, (int)_parent.GetLocation().PosY));
                    }
                }
                _elements.Clear();
                ResetComboDisplay();
                _reactionTimer = _maxTimer;
            }
        }
    }
}

File: aigilas/src/aigilas/reactions/ComboMeter.java
package aigilas.reactions;

import aigilas.creatures.BaseCreature;
import aigilas.entities.ComboMarker;
import aigilas.entities.Elements;
import sps.core.Logger;
import sps.text.ActionText;
import sps.text.TextManager;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class ComboMeter {
    private static final HashMap<Integer, Reaction> __reactions = new HashMap<Integer, Reaction>();

    static {
        __reactions.put(12, Reaction.Sweat);
        __reactions.put(13, Reaction.Magma);
        __reactions.put(14, Reaction.Explosion);
        __reactions.put(15, Reaction.Scorch);
        __reactions.put(16, Reaction.Blind);
        __reactions.put(17, Reaction.Lactic_Acid);
        __reactions.put(18, Reaction.Mind_Blown);
        __reactions.put(23, Reaction.Vent);
        __reactions.put(24, Reaction.Drown);
        __reactions.put(25, Reaction.Reflect);
        __reactions.put(26, Reaction.Drench);
        __reactions.put(27, Reaction.Pneumonia);
        __reactions.put(28, Reaction.Lobotomy);
        __reactions.put(34, Reaction.Rust);
        __reactions.put(35, Reaction.Purify);
        __reactions.put(36, Reaction.Eclipse);
        __reactions.put(37, Reaction.Respect);
        __reactions.put(38, Reaction.Craftsman);
        __reactions.put(45, Reaction.Flash);
        __reactions.put(46, Reaction.Metabolism);
        __reactions.put(47, Reaction.Fast_Forward);
        __reactions.put(48, Reaction.Blank);
        __reactions.put(56, Reaction.Yin_Yang);
        __reactions.put(57, Reaction.Expose);
        __reactions.put(58, Reaction.Enlighten);
        __reactions.put(67, Reaction.Atrophy);
        __reactions.put(68, Reaction.Neurosis);
        __reactions.put(78, Reaction.Confuse);
    }

    private final BaseCreature _parent;
    private final List<Elements> _elements = new ArrayList<Elements>();
    private final List<ComboMarker> _markers = new ArrayList<ComboMarker>();
    private static final int _maxTimer = 120;
    private int _reactionTimer = _maxTimer;

    public ComboMeter(BaseCreature parent) {
        _parent = parent;
    }

    private void resetComboDisplay() {
        _reactionTimer = _maxTimer;
        for (ComboMarker _marker : _markers) {
            _marker.setInactive();
        }
        _markers.clear();
        int jj = 0;
        for (Elements element : _elements) {
            _markers.add(new ComboMarker(_parent, element, jj));
            jj++;
            _markers.get(_markers.size() - 1).loadContent();
        }
    }

    public void draw() {
        for (ComboMarker marker : _markers) {
            marker.draw();
        }
    }

    public void add(Elements element) {
        if (!_elements.contains(element)) {
            if (_elements.size() == 2) {
                if (_elements.get(0).Value > element.Value) {
                    _elements.add(0, element);
                }
                else if (_elements.get(1).Value > element.Value) {
                    _elements.add(1, element);
                }
                else {
                    _elements.add(element);
                }
            }
            else if (_elements.size() == 1) {
                if (_elements.get(0).Value > element.Value) {
                    _elements.add(0, element);
                }
                else {
                    _elements.add(element);
                }
            }
            if (_elements.size() == 0) {
                _elements.add(element);
            }
            resetComboDisplay();
        }
    }

    private BaseReaction reaction;

    public void update() {
        for (ComboMarker marker : _markers) {
            marker.update();
        }
        int key;
        if (_elements.size() == 3) {
            key = _elements.get(0).Value * 100 + _elements.get(1).Value * 10 + _elements.get(2).Value;
            react(key);
        }
        if (_elements.size() == 2) {
            key = _elements.get(0).Value * 10 + _elements.get(1).Value;
            react(key);
        }
        if (_elements.size() == 1) {
            react(_elements.get(0).Value);
        }
    }

    private void react(int reactionId) {
        _reactionTimer--;
        if (_reactionTimer <= 0) {
            if (__reactions.keySet().contains(reactionId)) {
                reaction = ReactionFactory.create(__reactions.get(reactionId));
                if (reaction != null) {
                    reaction.affect(_parent);
                    Logger.gameplay(_parent + " affected by " + __reactions.get(reactionId).toString());
                    TextManager.add(new ActionText(__reactions.get(reactionId).toString(), 10, (int) _parent.getLocation().PosX, (int) _parent.getLocation().PosY));
                }
            }
            _elements.clear();
            resetComboDisplay();
            _reactionTimer = _maxTimer;
        }
    }
}
